### PR TITLE
unlynx_v2 the big and great cleanup

### LINKED ISSUE #11: Interface is nil
Sometimes and (randomly) I get an `interface is nil` error which I cannot replicate locally. Should try to get more information about it and fix it..

### PR DESCRIPTION
**Pull request by JoaoAndreSa**
I finished the big refactoring cleanup of Unlynx (let us call it v2 :P). I already merged it with master, but @ineiti if you see something that have missed or wrongly merged please let me know.

Since there are quite a lot of commits here is a list of the main changes:

**1. Refactor of packages:** I changed some of the packages: 
- Added a data package (contains the functions to generate fake data)
- The proofs in the lib package have been split into different packages (add_rm, aggregation, deterministic_tag...)
- no longer services/default package -> copied the files directly to services/

**2. Cleanup:** Deleted some unused code and added some more logs to catch all those naughty errors

**3. Proofs:** Homogenised and added hook functions for the proofs to define what to do with them (e.g. in unlynx we just create the proofs while in drynx we create them and send them to the skipchain).

**4. Removed the PARALLELIZE flag:** The parallelization is now by default.

**5. Added a new protocol:** shuffling+ddt protocol (does the shuffling and ddt in one single round)

**6. Fixed #11 :** just added the same `for el == nil {...}` in the HandleSurveyResponseQuery() to the HandleSurveyResultsQuery()

Please take a look at it and let me know if I should add or change things. This is the perfect time to do it (before I send it to master and create a new release).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,7 +6,7 @@ UnLynx is developed by lca1 (Laboratory for Communications and Applications in E
 
 ## Documentation
 
-* The UnLynx library does an intensive use of [Overlay-network (ONet) library](https://github.com/dedis/onet) and of the [Advanced Crypto (kyber) library](https://github.com/dedis/kyber).
+* The UnLynx library does an intensive use of [Overlay-network (ONet) library](https://go.dedis.ch/onet/v3onet) and of the [Advanced Crypto (kyber) library](https://go.dedis.ch/onet/v3kyber).
```
Comment (ineiti) [inline README.md]:
```suggestion
* The UnLynx library does an intensive use of [Overlay-network (ONet) library](https://github.com/dedis/onet) and of the [Advanced Crypto (kyber) library](https://github.com/dedis/kyber).
```

I think it's OK here to point to the actual github-repos.

File: app/client.go [hunk 2]
```diff
@@ -2,38 +2,37 @@ package appunlynx
...
-	surveyID, err := client.SendSurveyCreationQuery(el, servicesunlynxdefault.SurveyID(""), nil, nbrDPs, proofs, true, sum, count, whereQueryValues, predicate, groupBy)
+	surveyID, err := client.SendSurveyCreationQuery(el, servicesunlynx.SurveyID(""), nil, nbrDPs, proofs, true, sum, count, whereQueryValues, predicate, groupBy)
 	if err != nil {
 		log.Fatal("Service did not start.", err)
 	}
 
 	grp, aggr, err := client.SendSurveyResultsQuery(*surveyID)
 	if err != nil {
-		log.Fatal("Service could not output the results.")
+		log.Fatal("Service could not output the results.", err)
```
Comment (ineiti) [inline app/client.go]: Is this really `Fatal`, or should this just return the error? `Fatal` really means: "the internal state of the conode is inconsistent and every further action might make it worse". While here I think it should only be: "oups, something went wrong. Let's discard what happened and wait for a new request."

Comment (ineiti) [reply app/client.go]: Same thing for all `Fatal`s in the rest of the code...

File: data/handle_data.go [hunk 3]
```diff
@@ -17,7 +17,7 @@ import (
 )
 
 // Groups identifies all different groups to be added to the test data file
-var Groups [][]int64
+//var Groups [][]int64
```
Comment (ineiti) [inline data/handle_data.go]: Style: remove commented out code sections.

File: data/handle_data.go [hunk 4]
```diff
@@ -177,11 +185,18 @@ func WriteDataToFile(filename string, testData map[string][]libunlynx.DpClearRes
 	}
 
 	writer := bufio.NewWriter(fileHandle)
+	if err := writer.Flush(); err != nil {
+		log.Fatal(err)
```
Comment (ineiti) [inline data/handle_data.go]: like above: it's much cleaner to return an error and let the caller decide what should happen...

**Comment by ineiti** [comment]
Another downside of huge PRs: I don't even know where to start ;) Is there something specific that you would like me to look at?

File: app/client.go (hunk 2 above)
Comment (JoaoAndreSa) [reply app/client.go]: I am checking all `log.Fatal`s now. I will add a commit after reviewing all of them

Comment (JoaoAndreSa) [reply app/client.go]: @ineiti just commited this 5ad4d7e it should fix this comment (just of the app part)

File: data/handle_data.go (hunk 3 above)
Comment (JoaoAndreSa) [reply data/handle_data.go]: Done in 780f203

**Comment by JoaoAndreSa** [comment]
Apparently the previous bug in issue #11 was not fixed. I changed the service so that it waits for all the nodes to get the survey before performing the remaining operations/protocols: 083a047. 

@ineiti you told me something about a synchronous broadcast that waits until all nodes have gotten the message, but I only found something of that kind for the protocols: (*TreeNodeInstance) Broadcast(). I did the same with a channel, but I am not sure it's the best way (a protocols is perhaps the best choice, even though I would have to change some structures to avoid import cycles).

**Comment by ineiti** [comment]
In fact it's the `cothority/messaging/propagate` protocol. Unfortunately the test sucks and doesn't show how to do it nicely. But the skipchain uses it - follow `propagateGenesis` in here:

https://github.com/dedis/cothority/blob/master/skipchain/skipchain.go

**Pull request merged**